### PR TITLE
React Native App Start instrumentation

### PR DESCRIPTION
**Pull request by yousif-bugsnag**
## Goal

Adds App Start auto-instrumentation for React Native

## Design

If `autoInstrumentAppStarts` is true, `AppRegistry.setWrapperComponentProvider` is used to wrap the app components in a wrapper component that ends the app start span in a `useEffect` hook.

## Changeset

- Added a new `autoInstrumentAppStarts` config option
- Added a new `AppStartPlugin` - the plugin starts an app start span when the client is started and configuration is read, but using a start time recorded when the client is created (i.e. when `BugsnagPerformance` is imported)

## Testing

In the e2e tests scenarios are written as 'root' app components; the test fixture now starts Bugsnag with configuration defined by the scenario before calling `AppRegistry.registerComponent` and `AppRegistry.runApplication` to reload the app with the scenario's App component at the root.

### CODE REVIEW

File: test/react-native/features/steps/react-native-steps.rb [hunk 1]
```diff
@@ -7,7 +7,12 @@ def execute_command(action, scenario_name = '')
                 'local:9339'
               end
             else
-              'bs-local.com:9339'
+              case Maze::Helper.get_current_platform
```
Comment (yousif-bugsnag) [inline test/react-native/features/steps/react-native-steps.rb:10]: this fixes local testing on BrowserStack

File: test/react-native/features/support/env.rb [hunk 2]
```diff
@@ -1,4 +1,6 @@
 BeforeAll do
-  Maze.config.android_app_files_directory = '/data/local/tmp'
+  if Maze.config.farm == :bb
```
Comment (yousif-bugsnag) [inline test/react-native/features/support/env.rb:2]: BrowserStack doesn't allow pushing files to `/data/local/tmp` so we only override this on BitBar

File: packages/platforms/react-native/lib/auto-instrumentation/app-start-plugin.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  constructor (
+    spanFactory: SpanFactory<ReactNativeConfiguration>,
+    clock: Clock,
+    appRegistry: typeof AppRegistry
+  ) {
+    this.spanFactory = spanFactory
+    this.clock = clock
+    this.startTime = clock.now()
+    this.appRegistry = appRegistry
```
Comment (yousif-bugsnag) [inline packages/platforms/react-native/lib/auto-instrumentation/app-start-plugin.tsx:29]: AppRegistry is injected as a dependency here to aid unit testing

File: .buildkite/react-native-pipeline.yml [hunk 4]
```diff
@@ -87,6 +87,9 @@ steps:
           - --fail-fast
           - --no-tunnel
           - --aws-public-ip
+    retry:
+      manual:
+        permit_on_passed: true
```
Comment (gingerbenw) [inline .buildkite/react-native-pipeline.yml:92]: used to ensure reliability of flakey tests

File: packages/platforms/react-native/lib/auto-instrumentation/app-start-plugin.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        this.spanFactory.endSpan(appStartSpan, this.clock.now())
+      }, [])
+      return (<>{children}</>)
+    }
+
+    const instrumentedComponentProvider: WrapperComponentProvider = () => {
+      return ({children}) => {
+        return (<Wrapper>{children}</Wrapper>)
+      }
+    }
```
Comment (gingerbenw) [inline packages/platforms/react-native/lib/auto-instrumentation/app-start-plugin.tsx:50]:
```suggestion
    const instrumentedComponentProvider: WrapperComponentProvider = () => ({ children }) => {
        return (<Wrapper>{children}</Wrapper>)
    }
```

File: packages/platforms/react-native/lib/config.ts [hunk 6]
```diff
@@ -36,6 +38,11 @@ function createSchema (): ReactNativeSchema {
       defaultValue: true,
       message: 'should be true|false',
       validate: isBoolean
+    },
+    autoInstrumentAppStarts: {
+      defaultValue: true,
+      message: 'should be a boolean',
```
Comment (gingerbenw) [inline packages/platforms/react-native/lib/config.ts:44]:
```suggestion
      message: 'should be a true|false',
```

File: packages/platforms/react-native/tests/auto-instrumentation/app-start-plugin.test.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  beforeEach(() => {
+    spanFactory = new MockSpanFactory()
+    clock = createClock(performance)
+    appRegistry = {
+      setWrapperComponentProvider: jest.fn()
+    } as unknown as typeof AppRegistry
+  })
+
+  it('starts an app start span when autInstrumentAppStarts is true', () => {
```
Comment (gingerbenw) [inline packages/platforms/react-native/tests/auto-instrumentation/app-start-plugin.test.ts:21]:
```suggestion
  it('starts an app start span when autoInstrumentAppStarts is true', () => {
```

File: packages/platforms/react-native/tests/auto-instrumentation/app-start-plugin.test.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    expect(spanFactory.startSpan).toHaveBeenCalledWith('[AppStart/ReactNativeInit]',
+      expect.objectContaining({
+        startTime: expect.any(Number),
+        parentContext: null
+      }))
+
+    expect(appRegistry.setWrapperComponentProvider).toHaveBeenCalledWith(expect.any(Function))
+  })
+
+  it('does not start an app start span when autInstrumentAppStarts is false', () => {
```
Comment (gingerbenw) [inline packages/platforms/react-native/tests/auto-instrumentation/app-start-plugin.test.ts:35]:
```suggestion
  it('does not start an app start span when autoInstrumentAppStarts is false', () => {
```

File: test/react-native/features/fixtures/app/babel.config.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  presets: ['@babel/preset-env', 'module:metro-react-native-babel-preset'],
+  plugins: [['module:react-native-dotenv']]
```
Comment (gingerbenw) [inline test/react-native/features/fixtures/app/babel.config.js:3]: Don't think this needs to be an array item if there are no options provided

```suggestion
  plugins: ['module:react-native-dotenv']
```

File: test/react-native/features/fixtures/performance-scenarios/scenarios/ManualSpanScenario.js [hunk 10]
```diff
@@ -2,25 +2,24 @@ import React, { useEffect } from 'react'
...
-  const App = () => {
-    useEffect(() => {
-      BugsnagPerformance.start({ apiKey, endpoint, maximumBatchSize: 1, appName: 'com.bugsnag.reactnative.performance', appVersion: '1.2.3' })
-      BugsnagPerformance.startSpan('ManualSpanScenario').end()
-    }, [])
+export const config = {
+  maximumBatchSize: 1,
+  autoInstrumentAppStarts: false,
+  appVersion: '1.2.3'
+}
```
Comment (gingerbenw) [inline test/react-native/features/fixtures/performance-scenarios/scenarios/ManualSpanScenario.js:9]: definitely neater - but are you still starting BugsnagPerformance?

**Review by gingerbenw — CHANGES_REQUESTED** [review]

File: packages/platforms/react-native/lib/auto-instrumentation/app-start-plugin.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  private readonly appRegistry: typeof AppRegistry
+
+  constructor (
+    spanFactory: SpanFactory<ReactNativeConfiguration>,
+    clock: Clock,
+    appRegistry: typeof AppRegistry
+  ) {
+    this.spanFactory = spanFactory
+    this.clock = clock
+    this.startTime = clock.now()
```
[the remark this answers is not in the document]
Comment (yousif-bugsnag) [reply packages/platforms/react-native/lib/auto-instrumentation/app-start-plugin.tsx:28]: yes, that's a good point

File: test/react-native/features/fixtures/performance-scenarios/scenarios/ManualSpanScenario.js (hunk 10 above)
Comment (yousif-bugsnag) [reply test/react-native/features/fixtures/performance-scenarios/scenarios/ManualSpanScenario.js:9]: yep - BugsnagPerformance is now started in `ScenarioLauncher.js`, before the scenario is loaded.

this allows us to better test the app start instrumentation as the call to start sets a wrapper component provider which then wraps our scenario component

File: test/react-native/features/fixtures/scenario-launcher/lib/ScenarioLauncher.js [hunk 12]
```diff
@@ -1,17 +1,30 @@
 import { AppRegistry } from 'react-native'
 import { getCurrentCommand } from './CommandRunner'
-import * as Scenarios from './scenarios/Scenarios'
+import * as Scenarios from '../scenarios'
 import { REACT_APP_SCENARIO_NAME, REACT_APP_ENDPOINT, REACT_APP_API_KEY } from '@env'
+import BugsnagPerformance from '@bugsnag/react-native-performance'
+
+const delay = ms => new Promise(resolve => setTimeout(resolve, ms))
```
Comment (gingerbenw) [inline test/react-native/features/fixtures/scenario-launcher/lib/ScenarioLauncher.js:7]: unused function

**Review by gingerbenw — APPROVED** [review]

File: test/react-native/features/fixtures/scenario-launcher/lib/ScenarioLauncher.js [hunk 13]
```diff
@@ -1,17 +1,30 @@
...
```
Comment (gingerbenw) [inline test/react-native/features/fixtures/scenario-launcher/lib/ScenarioLauncher.js:8]:
```suggestion
```

**Pull request merged**